core-bundle/src/Command/MigrateCommand.php

```diff
@@ -0,0 +1,272 @@
+        $this->installer = $installer;
+
+        parent::__construct();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function configure(): void
+    {
+        $this
+            ->setName('contao:migrate')
+            ->addOption('complete', null, InputOption::VALUE_NONE, 'Execute all database migrations including DROP queries. Can be used together with --no-interaction.')
+            ->addOption('schema-only', null, InputOption::VALUE_NONE, 'Execute database schema migration only.')
+            ->setDescription('Executes migrations and the database schema diff.')
+        ;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $this->io = new SymfonyStyle($input, $output);
+
+        if ($input->getOption('schema-only')) {
+            return $this->executeSchemaDiff($input->getOption('complete')) ? 0 : 1;
+        }
+
+        if (!$this->executeMigrations()) {
+            return 1;
+        }
+
+        if (!$this->executeSchemaDiff($input->getOption('complete'))) {
+            return 1;
+        }
+
+        if (!$this->executeMigrations()) {
+            return 1;
+        }
+
+        $this->io->success('All migrations completed.');
+
+        return 0;
+    }
+
+    private function executeMigrations(): bool
+    {
+        while (true) {
+            $first = true;
+
+            foreach ($this->migrations->getPendingMigrations() as $migration) {
+                if ($first) {
+                    $this->io->section('Pending migrations');
+                    $first = false;
+                }
+
+                $this->io->writeln(' * '.$migration);
+            }
+
+            foreach ($this->getRunOnceFiles() as $file) {
+                if ($first) {
+                    $this->io->section('Pending migrations');
+                    $first = false;
+                }
+
+                $this->io->writeln(' * Runonce file: '.$file);
+            }
+
+            if ($first) {
+                break;
+            }
+
+            if (!$this->io->confirm('Execute the listed migrations?')) {
+                return false;
+            }
+
+            $this->io->section('Execute migrations');
+
+            foreach ($this->migrations->runMigrations() as $result) {
+                $this->io->writeln(' * '.$result->getMessage());
+
+                if (!$result->isSuccessful()) {
+                    $this->io->error('Migration failed');
+                }
+            }
+
+            foreach ($this->getRunOnceFiles() as $file) {
+                $this->executeRunonceFile($file);
+                $this->io->writeln(' * Executed runonce file: '.$file);
+            }
+
+            $this->io->success('Executed migrations.');
+        }
+
+        return true;
+    }
+
+    private function getRunOnceFiles(): array
+    {
+        try {
+            $files = $this->fileLocator->locate('config/runonce.php', null, false);
```

Comment:
Do you think we should add a warning/deprecation message so users can inform the developers about our new migrations framework? 🎉

Reply:
Agree

Reply:
ec79cc14865c923fa1f1dae83b31e1d2ed6ea995